### PR TITLE
[V2] chore: Move SharedState struct and related functions to shared_state.go

### LINKED ISSUE #1412: [V2] chore: Move SharedState struct and related functions to shared_state.go
**Is your feature request related to a problem?/Why is this needed**

Currently, the `SharedState` struct and its related functions are scattered through source files in the `pkg/controller` package. This makes maintaining them harder.

**Describe the solution you'd like in detail**

1. Move the `SharedState` struct and its related functions into a single source file, `shared_state.go`
2. Make the `SharedState` struct package private, if possible.
3. Add `SharedState` focused unit tests.
4. Consider making `SharedState` and anonymous field in the controllers to avoid the need to explicitly access if by field name (currently, `controllerSharedState` which is redundant information and really long to type).

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by shlokshah-dev**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Currently, the SharedState struct and its related functions are scattered through source files in the pkg/controller package. This makes maintaining them harder. This PR refactors the SharedState struct and moves it and its related method to a separate file, making it easier to maintain.

**Which issue(s) this PR fixes**:

Fixes #1412

**Requirements**:
- [X] Move the `SharedState` struct and its related functions into a single source file, `shared_state.go`
- [X] Make the `SharedState` struct package private, if possible.
- [X] Consider making `SharedState` and anonymous field in the controllers to avoid the need to explicitly access if by field name (currently, controllerSharedState which is redundant information and really long to type).

**Special notes for your reviewer**:
Adding SharedState specific unit tests will be taken care as part of a separate PR.

**Release note**:
```
none
```

### CODE REVIEW

**Review by sunpa93 — CHANGES_REQUESTED** [review]
Seems like we are missing changes from https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1452, which will have some conflicts with your changes. You will need to rebase to apply your commit on top of existing changes.

File: pkg/controller/attach_detach.go [hunk 1]
```diff
@@ -63,12 +63,12 @@ Attach Detach controller is responsible for
...
-	cloudDiskAttacher     CloudDiskAttachDetacher
-	controllerSharedState *SharedState
-	stateLock             *sync.Map
-	retryInfo             *retryInfo
+	logger            logr.Logger
+	crdDetacher       CrdDetacher
+	cloudDiskAttacher CloudDiskAttachDetacher
+	stateLock         *sync.Map
+	retryInfo         *retryInfo
+	*sharedState
```
Comment (edreed) [inline pkg/controller/attach_detach.go:71]: Place anonymous fields at the beginning rather than the end.

File: pkg/controller/common.go [hunk 2]
```diff
@@ -42,20 +40,21 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/component-helpers/scheduling/corev1/nodeaffinity"
 	azdiskv1beta2 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta2"
-	azdisk "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
-	azdiskinformers "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/informers/externalversions"
+
+	// azdisk "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
```
Comment (sunpa93) [inline pkg/controller/common.go:44]: np: Please remove these commented lines.

File: pkg/controller/shared_state.go [hunk 3]
```diff
@@ -0,0 +1,1151 @@
...
+	azdisk "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	azdiskinformers "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/informers/externalversions"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/watcher"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/workflow"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type sharedState struct {
```
Comment (sunpa93) [inline pkg/controller/shared_state.go:50]: This needs to be public as it is used in `azuredisk` pkg

https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/7f34b3cbcefed45b23a12a610d9bf5b034700dc4/pkg/azuredisk/azuredisk_v2.go#L387

File: pkg/controller/shared_state.go [hunk 4]
```diff
@@ -0,0 +1,1151 @@
...
+	processingReplicaRequestQueue int32
+	eventRecorder                 record.EventRecorder
+	cachedClient                  client.Client
+	azClient                      azdisk.Interface
+	kubeClient                    kubernetes.Interface
+	conditionWatcher              *watcher.ConditionWatcher
+}
+
+func NewSharedState(driverName, objectNamespace, topologyKey string, eventRecorder record.EventRecorder, cachedClient client.Client, azClient azdisk.Interface, kubeClient kubernetes.Interface) *sharedState {
+	newSharedState := &sharedState{driverName: driverName, objectNamespace: objectNamespace, topologyKey: topologyKey, eventRecorder: eventRecorder, cachedClient: cachedClient, azClient: azClient, kubeClient: kubeClient, conditionWatcher: watcher.New(context.Background(), azClient, azdiskinformers.NewSharedInformerFactory(azClient, consts.DefaultInformerResync), objectNamespace)}
```
Comment (edreed) [inline pkg/controller/shared_state.go:76]: Please place field assignments on separate lines.

File: pkg/controller/attach_detach.go (hunk 1 above)
Comment (shlokshah-dev) [reply pkg/controller/attach_detach.go]: Done.

File: pkg/controller/common.go (hunk 2 above)
Comment (shlokshah-dev) [reply pkg/controller/common.go]: Done.

File: pkg/controller/shared_state.go (hunk 3 above)
Comment (shlokshah-dev) [reply pkg/controller/shared_state.go]: Updated.

File: pkg/controller/shared_state.go (hunk 4 above)
Comment (shlokshah-dev) [reply pkg/controller/shared_state.go]: Done.

**Pull request merged**